File: solutions/ts/05.ts
type Shoe = {
  type: "I" | "R";
  size: number;
};
type SizeItem = {
  I: number;
  R: number;
  size: number;
};

function organizeShoes(shoes: Shoe[]): number[] {
  const sizes: Record<number, SizeItem> = {};
  const result: number[] = [];
  for (const shoe of shoes) {
    sizes[shoe.size] ??= { I: 0, R: 0, size: shoe.size };
    sizes[shoe.size][shoe.type] += 1;
  }
  for (const item of Object.values(sizes)) {
    while (item.I > 0 && item.R > 0) {
      item.I--;
      item.R--;
      result.push(item.size);
    }
  }
  return result;
}

const shoes: Shoe[] = [
  { type: "I", size: 38 },
  { type: "R", size: 38 },
  { type: "R", size: 42 },
  { type: "I", size: 41 },
  { type: "I", size: 42 },
];
const pairs1 = organizeShoes(shoes);
console.log(pairs1);
// [38, 42]

const shoes2: Shoe[] = [
  { type: "I", size: 38 },
  { type: "R", size: 38 },
  { type: "I", size: 38 },
  { type: "I", size: 38 },
  { type: "R", size: 38 },
];
const pairs2 = organizeShoes(shoes2);
console.log(pairs2);
// [38, 38]

const shoes3: Shoe[] = [
  { type: "I", size: 38 },
  { type: "R", size: 36 },
  { type: "R", size: 42 },
  { type: "I", size: 41 },
  { type: "I", size: 42 },
];
const pairs3 = organizeShoes(shoes3);
console.log(pairs3);
// [42]

/* 
  Score: ★★★★★
*/

File: solutions/python/05.py
def organizeShoes(shoes):
    sizes = {}
    result = []
    for shoe in shoes:
        sizes.setdefault(shoe["size"], {"I": 0, "R": 0, "size": shoe["size"]})
        sizes[shoe["size"]][shoe["type"]] += 1
    for item in sizes.values():
        while (item["I"] > 0 and item["R"] > 0):
            item["I"] -= 1
            item["R"] -= 1
            result.append(item["size"])
    return result


if __name__ == "__main__":
    shoes = [
        {"type": "I", "size": 38},
        {"type": "R", "size": 38},
        {"type": "R", "size": 42},
        {"type": "I", "size": 41},
        {"type": "I", "size": 42},
    ]
    result1 = organizeShoes(shoes)
    print(result1)
    # [38, 42]

    shoes2 = [
        {"type": "I", "size": 38},
        {"type": "R", "size": 38},
        {"type": "I", "size": 38},
        {"type": "I", "size": 38},
        {"type": "R", "size": 38},
    ]
    result2 = organizeShoes(shoes2)
    print(result2)
    # [38, 38]

    shoes3 = [
        {"type": "I", "size": 38},
        {"type": "R", "size": 36},
        {"type": "R", "size": 42},
        {"type": "I", "size": 41},
        {"type": "I", "size": 42},
    ]
    result3 = organizeShoes(shoes3)
    print(result3)
    # [42]

"""
  Score: ★★★★★
"""
